llvm/lib/SYCLLowerIR/AddAspectNames.cpp

```diff
@@ -0,0 +1,69 @@
+//===--------------- AddAspectNames.cpp - AddAspectNames Pass -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// The !sycl_used_aspects metadata is populated from C++ attributes and
+// further populated by the SYCLPropagateAspectsPass that describes which
+// apsects a function uses. The format of this metadata initially just an
+// integer value corresponding to the enum value in C++. The !sycl_aspects
+// named metadata contains the associations from aspect values to aspect names.
+// This pass takes the associations from !sycl_aspects and then updates all the
+// !sycl_used_aspects metadata to include the aspect names.
+//===----------------------------------------------------------------------===//
+//
+
+#include "llvm/SYCLLowerIR/AddAspectNames.h"
+
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Module.h"
+
+using namespace llvm;
+
+PreservedAnalyses AddAspectNamesPass::run(Module &M,
+                                          ModuleAnalysisManager &MAM) {
+  SmallDenseMap<int64_t, Metadata *, 128> ValueToNameValuePairMD;
+  if (NamedMDNode *Node = M.getNamedMetadata("sycl_aspects")) {
+    for (MDNode *N : Node->operands()) {
+      assert(N->getNumOperands() == 2 &&
+             "Each operand of sycl_aspects must be a pair.");
+
+      // The aspect's integral value is the second operand.
+      const auto *AspectCAM = cast<ConstantAsMetadata>(N->getOperand(1));
+      const Constant *AspectC = AspectCAM->getValue();
+
+      ValueToNameValuePairMD[cast<ConstantInt>(AspectC)->getSExtValue()] = N;
```

Comment:
```suggestion
      auto *C = mdconst::extract<ConstantInt>(N->getOperand(1));

      ValueToNameValuePairMD[C->getSExtValue()] = N;
```

Documentation: [`mdconst` namespace](https://llvm.org/doxygen/namespacellvm_1_1mdconst.html)